cabal-install/Distribution/Client/Glob.hs

```diff
@@ -54,53 +161,116 @@ globMatches (Glob atoms) = goStart atoms
       | otherwise               = False
     go [WildCard]         ""    = True
     go (WildCard:rest)   (c:cs) = go rest (c:cs) || go (WildCard:rest) cs
-    go (Union globs:rest)   cs  = any (\(Glob glob) ->
-                                          go (glob ++ rest) cs) globs
+    go (Union globs:rest)   cs  = any (\glob -> go (glob ++ rest) cs) globs
     go []                (_:_)  = False
     go (_:_)              ""    = False
 
-instance Text Glob where
-  disp (Glob atoms) = Disp.hcat (map dispAtom atoms) 
-    where
-      dispAtom WildCard      = Disp.char '*'
-      dispAtom (Literal str) = Disp.text (escape str)
-      dispAtom (Union globs) = Disp.braces
-                                 (Disp.hcat (Disp.punctuate (Disp.char ',')
-                                                            (map disp globs)))
-
-      escape []               = []
-      escape (c:cs)
-        | isGlobEscapedChar c = '\\' : c : escape cs
-        | otherwise           =        c : escape cs
-
-  parse = Glob `fmap` many1 globAtom
+
+------------------------------------------------------------------------------
+-- Parsing & printing
+--
+
+instance Text FilePathGlob where
+  disp (FilePathGlob root pathglob) = disp root Disp.<> disp pathglob
+  parse =
+    parse >>= \root ->
+        (FilePathGlob root <$> parse)
+    <++ (when (root == FilePathRelative) Parse.pfail >>
+         return (FilePathGlob root GlobDirTrailing))
+
+instance Text FilePathRoot where
+  disp  FilePathRelative    = Disp.empty
+  disp  FilePathUnixRoot    = Disp.char '/'
+  disp (FilePathWinDrive c) = Disp.char c
+                      Disp.<> Disp.char ':'
+                      Disp.<> Disp.char '\\'
+  disp FilePathHomeDir      = Disp.char '~'
+                      Disp.<> Disp.char '/'
+
+  parse =
+        (     (Parse.char '/' >> return FilePathUnixRoot)
+          +++ (Parse.char '~' >> Parse.char '/' >> return FilePathHomeDir)
+          +++ (do drive <- Parse.satisfy (\c -> (c >= 'a' && c <= 'z')
```

Comment:
We may also want to support UNC paths (that is, `\\SERVERNAME\user\path\to\file`).